torch/nn/functional.py

```diff
@@ -3829,26 +3829,125 @@ def fold(input, output_size, kernel_size, dilation=1, padding=0, stride=1):
 def _pad_circular(input, padding):
     # type: (Tensor, List[int]) -> Tensor
     """
-    Arguments
-        :param input: tensor of shape :math:`(N, C_{\text{in}}, H, [W, D]))`
-        :param padding: (tuple): m-elem tuple where m is the degree of convolution
-    Returns
-        :return: tensor of shape :math:`(N, C_{\text{in}}, [D + 2 * padding[0],
-                 H + 2 * padding[1]], W + 2 * padding[2]))`
+    Args:
+        input: Tensor that follows the formatting of the input to convolution
+            layers.
+        padding: Tuple with length two times the degree of the convolution. The
+            order of the integers in the tuple are shown in the following
+            example:
+
+            For 3D convolutions:
+                padding[-2] is the amount of padding applied to the beginning
+                    of the depth dimension.
+                padding[-1] is the amount of padding applied to the end of the
+                    depth dimension.
+                padding[-4] is the amount of padding applied to the beginning
+                    of the height dimension.
+                padding[-3] is the amount of padding applied to the end of the
+                    height dimension.
+                padding[-6] is the amount of padding applied to the beginning
+                    of the width dimension.
+                padding[-5] is the amount of padding applied to the end of the
+                    width dimension.
+
+    Returns:
+        out: Tensor with padded shape.
     """
+    shape = input.shape
+    ndim = len(shape[2:])
 
-    input = torch.cat([input, input[:, :, 0:padding[-1]]], dim=2)
-    input = torch.cat([input[:, :, -(padding[-1] + padding[-2]):-padding[-1]], input], dim=2)
+    # Only supports wrapping around once
+    for a, size in enumerate(shape[2:]):
```

Comment:
A test case will need to be added to validate these asserts, too.

Reply:
Here are some cases that validate these asserts, as well as the new assert for negative padding.
```python
x = torch.randn((1, 2, 12, 36, 48), dtype=torch.float32, device=device)
padding = (48, 48, 36, 36, 12, 12)
padding = (48, 49, 36, 36, 12, 12)  # Raises assertion error
padding = (48, 48, 36, 36, 13, 12)  # Raises assertion error
padding = (0, 0, 0, 0, -7, -5)
padding = (0, 0, 0, 0, -7, -6)  # Raises assertion error
y = _pad_circular(x, padding)
```